--- a/net/minecraft/world/level/block/entity/TestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TestBlockEntity.java
@@ -38,6 +_,7 @@
 
     @Override
     protected void loadAdditional(ValueInput valueInput) {
+        super.loadAdditional(valueInput); // Paper - load the PDC
         this.mode = valueInput.read("mode", TestBlockMode.CODEC).orElse(TestBlockMode.FAIL);
         this.message = valueInput.getStringOr("message", "");
         this.powered = valueInput.getBooleanOr("powered", false);
