--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -82,6 +_,7 @@
 
             for (Leashable leashable : Leashable.leashableLeashedTo(player)) {
                 if (leashable.canHaveALeashAttachedTo(this)) {
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(leashable, this, player, hand, false)) continue; // CraftBukkit - Call PlayerLeashEntityEvent
                     leashable.setLeashedTo(this, true);
                     flag = true;
                 }
@@ -91,6 +_,7 @@
             if (!flag && !player.isSecondaryUseActive()) {
                 for (Leashable leashable1 : Leashable.leashableLeashedTo(this)) {
                     if (leashable1.canHaveALeashAttachedTo(player)) {
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(leashable1, player, player, hand, false)) continue; // CraftBukkit - call PlayerLeashEntityEvent
                         leashable1.setLeashedTo(player, true);
                         flag1 = true;
                     }
