--- a/net/minecraft/world/entity/AgeableMob.java
+++ b/net/minecraft/world/entity/AgeableMob.java
@@ -23,6 +_,7 @@
     protected int age = 0;
     protected int forcedAge = 0;
     protected int forcedAgeTimer;
+    public boolean ageLocked; // CraftBukkit
 
     protected AgeableMob(EntityType<? extends AgeableMob> entityType, Level level) {
         super(entityType, level);
@@ -69,13 +_,15 @@
     }
 
     public void ageUp(int amount, boolean forced) {
+        if (this.ageLocked) return; // Paper - Honor ageLock
         int age = this.getAge();
+        int previousAge = age; // Paper - Decompile fix: lvt reassignment lost
         age += amount * 20;
         if (age > 0) {
             age = 0;
         }
 
-        int i1 = age - age;
+        int i1 = age - previousAge; // Paper - Decompile fix
         this.setAge(age);
         if (forced) {
             this.forcedAge += i1;
@@ -107,6 +_,7 @@
         super.addAdditionalSaveData(valueOutput);
         valueOutput.putInt("Age", this.getAge());
         valueOutput.putInt("ForcedAge", this.forcedAge);
+        valueOutput.putBoolean("AgeLocked", this.ageLocked); // CraftBukkit
     }
 
     @Override
@@ -114,6 +_,7 @@
         super.readAdditionalSaveData(valueInput);
         this.setAge(valueInput.getIntOr("Age", 0));
         this.forcedAge = valueInput.getIntOr("ForcedAge", 0);
+        this.ageLocked = valueInput.getBooleanOr("AgeLocked", false); // CraftBukkit
     }
 
     @Override
@@ -128,7 +_,7 @@
     @Override
     public void aiStep() {
         super.aiStep();
-        if (this.level().isClientSide) {
+        if (this.level().isClientSide || this.ageLocked) { // CraftBukkit
             if (this.forcedAgeTimer > 0) {
                 if (this.forcedAgeTimer % 4 == 0) {
                     this.level().addParticle(ParticleTypes.HAPPY_VILLAGER, this.getRandomX(1.0), this.getRandomY() + 0.5, this.getRandomZ(1.0), 0.0, 0.0, 0.0);
