From a4974fec545b37826e26e3cfe35b78eed00600ef Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/server/bossevents/CustomBossEvent.java b/net/minecraft/server/bossevents/CustomBossEvent.java
index 0a95a92c72efab798a8cecd16cbbfb67ffe043e7..8edaed10ad46055c9a06e717545c758e69be5a62 100644
--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -23,6 +23,16 @@ public class CustomBossEvent extends ServerBossEvent {
     private final Set<UUID> players = Sets.newHashSet();
     private int value;
     private int max = 100;
+    // CraftBukkit start
+    private org.bukkit.boss.KeyedBossBar bossBar;
+
+    public org.bukkit.boss.KeyedBossBar getBukkitEntity() {
+        if (this.bossBar == null) {
+            this.bossBar = new org.bukkit.craftbukkit.boss.CraftKeyedBossbar(this);
+        }
+        return this.bossBar;
+    }
+    // CraftBukkit end
 
     public CustomBossEvent(ResourceLocation id, Component name) {
         super(name, BossEvent.BossBarColor.WHITE, BossEvent.BossBarOverlay.PROGRESS);
