From a4974fec545b37826e26e3cfe35b78eed00600ef Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
index d2c0ae2abd1e16996e7d3bafa4cb6c4004ea337e..4fb63e58eac5d67fcd31c3233dca1dae72b98bc4 100644
--- a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -111,11 +111,17 @@ public class VillagerMakeLove extends Behavior<Villager> {
         if (breedOffspring == null) {
             return Optional.empty();
         } else {
-            parent.setAge(6000);
-            partner.setAge(6000);
             breedOffspring.setAge(-24000);
             breedOffspring.moveTo(parent.getX(), parent.getY(), parent.getZ(), 0.0F, 0.0F);
-            level.addFreshEntityWithPassengers(breedOffspring);
+            // CraftBukkit start - call EntityBreedEvent
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreedEvent(breedOffspring, parent, partner, null, null, 0).isCancelled()) {
+                return Optional.empty();
+            }
+            // Move age setting down
+            parent.setAge(6000);
+            partner.setAge(6000);
+            level.addFreshEntityWithPassengers(breedOffspring, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING);
+            // CraftBukkit end
             level.broadcastEntityEvent(breedOffspring, (byte)12);
             return Optional.of(breedOffspring);
         }
